pkg/querier/querier.go

```diff
@@ -97,30 +101,37 @@ func New(cfg Config, distributor Distributor, chunkStore ChunkStore) (storage.Qu
 }
 
 // NewQueryable creates a new Queryable for cortex.
-func NewQueryable(dq, cq storage.Queryable, distributor Distributor, ingesterMaxQueryLookback time.Duration) storage.Queryable {
+func NewQueryable(dq, cq storage.Queryable, distributor Distributor, cfg Config) storage.Queryable {
 	return storage.QueryableFunc(func(ctx context.Context, mint, maxt int64) (storage.Querier, error) {
-		cqr, err := cq.Querier(ctx, mint, maxt)
-		if err != nil {
-			return nil, err
-		}
-
 		q := querier{
-			queriers:    []storage.Querier{cqr},
 			distributor: distributor,
 			ctx:         ctx,
 			mint:        mint,
 			maxt:        maxt,
 		}
 
 		// Include ingester only if maxt is within ingesterMaxQueryLookback w.r.t. current time.
-		if ingesterMaxQueryLookback == 0 || maxt >= time.Now().Add(-ingesterMaxQueryLookback).UnixNano()/1e6 {
+		now := model.Now()
+		if cfg.IngesterMaxQueryLookback == 0 || maxt >= int64(now.Add(-cfg.IngesterMaxQueryLookback)) {
 			dqr, err := dq.Querier(ctx, mint, maxt)
 			if err != nil {
 				return nil, err
 			}
 			q.queriers = append(q.queriers, dqr)
+
+			// metrics wont have been sent to store yet, skip store query
+			if model.Time(mint).After(now.Add(-cfg.IngesterMetricMinAge)) {
```

Comment:
I think having this `if` here makes the code a bit difficult to read, because the parent `if` block is used to determine if a query should hit the ingesters, while this `if` block is used to determine if a query should hit the store.

Instead of having a guard here, I would suggest to wrap the store querier within a `if`, to keep the intention more explicit.

_Same comment applies to `pkg/querier/unified_querier.go`_

Reply:
Good call, I like that better too.